### ISSUE TITLE
h2 cancellations causing connection leak / file descriptor leak

### ISSUE THREAD

**Issue by klingerf** [linkerd/linkerd#1342]
First reported by @zackangelo.

Am running a linker-to-linker daemonset setup in Kubernetes, routing gRPC traffic via h2 routers. The gRPC client has an aggressive timeout which the backend server can't meet most of the time. This results in the client sending a RST_STREAM frame to cancel the request.

For a given stream (71, in this example), when the client cancels I see a bunch of these messages in Linkerd's logs:

```
I 0524 00:38:49.932 UTC THREAD41: [S L:/10.196.2.232:5757 R:/10.196.2.234:45676 S:71] interrupted; resetting remote: CANCEL
Failure(com.twitter.finagle.buoyant.h2.Reset$Cancel$, flags=0x03) with RemoteInfo -> Upstream Address: /10.196.2.234:45676, Upstream Client Id: Not Available, Downstream Address: 10.196.0.243/10.196.0.243:6767, Downstream Client Id: %/io.l5d.k8s.daemonset/test/incoming/l5d7/#/io.l5d.k8s/test/grpc/strest-server7, Trace Id: 480173199f5e724b.480173199f5e724b<:480173199f5e724b with Service -> 0.0.0.0/5757
Caused by: Reset.Cancel
```

Here's a full stats dump from one of my linkerds:

[conn-leak-metrics.txt](https://github.com/linkerd/linkerd/files/1024127/conn-leak-metrics.txt)

In my test, I'm only sending 25 concurrent requests to the outgoing router, but you can see that that has resulted in 321 open connections from the outgoing router to the incoming router:

```
  "rt/outgoing/client/%/io.l5d.k8s.daemonset/test/incoming/l5d7/#/io.l5d.k8s/test/grpc/strest-server7/connections": 321,
```

The connections continue to climb as more cancels are received. With an unlimited file descriptor ulimit, the process eventually runs out of memory. If the file descriptor limit is bounded, then the process eventually throws a file descriptor error:

```
WARN 0524 00:38:53.640 UTC finagle/netty4-3: Failed to initialize a channel. Closing: [id: 0xdcd02305, L:/10.196.2.232:6767 - R:/10.196.2.232:36216]
java.lang.IllegalArgumentException: File does not contain valid certificates: /io.buoyant/linkerd/certs/linkerdcombined.crt
	at io.netty.handler.ssl.SslContextBuilder.keyManager(SslContextBuilder.java:247)
	at io.netty.handler.ssl.SslContextBuilder.keyManager(SslContextBuilder.java:207)
	at io.netty.handler.ssl.SslContextBuilder.forServer(SslContextBuilder.java:50)
	at com.twitter.finagle.netty4.ssl.server.Netty4ServerEngineFactory.startWithKey(Netty4ServerEngineFactory.scala:23)
	at com.twitter.finagle.netty4.ssl.server.Netty4ServerEngineFactory.apply(Netty4ServerEngineFactory.scala:45)
	at com.twitter.finagle.netty4.ssl.server.Netty4ServerSslHandler.$anonfun$initChannel$1(Netty4ServerSslHandler.scala:70)
	at com.twitter.finagle.netty4.ssl.server.Netty4ServerSslHandler.$anonfun$initChannel$1$adapted(Netty4ServerSslHandler.scala:67)
	at scala.Option.foreach(Option.scala:257)
	at com.twitter.finagle.netty4.ssl.server.Netty4ServerSslHandler.initChannel(Netty4ServerSslHandler.scala:67)
	at io.netty.channel.ChannelInitializer.initChannel(ChannelInitializer.java:113)
	at io.netty.channel.ChannelInitializer.handlerAdded(ChannelInitializer.java:105)
	at io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:597)
	at io.netty.channel.DefaultChannelPipeline.addFirst(DefaultChannelPipeline.java:178)
	at io.netty.channel.DefaultChannelPipeline.addFirst(DefaultChannelPipeline.java:143)
	at com.twitter.finagle.netty4.channel.Netty4RawServerChannelInitializer.initChannel(Netty4RawServerChannelInitializer.scala:55)
	at io.netty.channel.ChannelInitializer.initChannel(ChannelInitializer.java:113)
	at io.netty.channel.ChannelInitializer.handlerAdded(ChannelInitializer.java:105)
	at io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:597)
	at io.netty.channel.DefaultChannelPipeline.addLast(DefaultChannelPipeline.java:226)
	at io.netty.channel.DefaultChannelPipeline.addLast(DefaultChannelPipeline.java:392)
	at io.netty.channel.DefaultChannelPipeline.addLast(DefaultChannelPipeline.java:379)
	at com.twitter.finagle.netty4.Netty4Listener$$anon$1$$anon$2.initChannel(Netty4Listener.scala:167)
	at io.netty.channel.ChannelInitializer.initChannel(ChannelInitializer.java:113)
	at io.netty.channel.ChannelInitializer.handlerAdded(ChannelInitializer.java:105)
	at io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:597)
	at io.netty.channel.DefaultChannelPipeline.access$000(DefaultChannelPipeline.java:44)
	at io.netty.channel.DefaultChannelPipeline$PendingHandlerAddedTask.execute(DefaultChannelPipeline.java:1387)
	at io.netty.channel.DefaultChannelPipeline.callHandlerAddedForAllHandlers(DefaultChannelPipeline.java:1122)
	at io.netty.channel.DefaultChannelPipeline.invokeHandlerAddedIfNeeded(DefaultChannelPipeline.java:647)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:506)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:419)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:478)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:442)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at com.twitter.finagle.util.BlockingTimeTrackingThreadFactory$$anon$1.run(BlockingTimeTrackingThreadFactory.scala:24)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.security.cert.CertificateException: could not find certificate file: /io.buoyant/linkerd/certs/linkerdcombined.crt
	at io.netty.handler.ssl.PemReader.readCertificates(PemReader.java:69)
	at io.netty.handler.ssl.SslContext.toX509Certificates(SslContext.java:1026)
	at io.netty.handler.ssl.SslContextBuilder.keyManager(SslContextBuilder.java:245)
	... 39 more
```

In this case it's trying to load the SSL certificate file in order to establish a new connection, but the file fails to load and the process exits. It's surprising to me that the certificate file is loaded from disk each time a new connection is established.

**Comment by klingerf** [comment]
Each reset generates an `h2.Reset.Cancel` error, and Zack notes that based on the stats, it looks like those are being counted toward failure accrual, when they probably shouldn't be.

**Comment by adleong** [comment]
https://github.com/linkerd/linkerd/pull/1444 Should fix this issue

**Comment by klingerf** [comment]
Re-ran my initial test with the fix from #1444, and verified that it is no longer leaking connections. Thanks @adleong! Will close this.
